### PR TITLE
MAKES grammar reference and how-to guide

### PR DESCRIPTION
**Pull request by DarrinEide**

### CODE REVIEW

File: academic-services/knowledge-exploration-service/how-to-index-data.md [hunk 1]
```diff
@@ -44,7 +44,7 @@ KES allows this to be defined for an entity as a relative log probability using
 An entity data file contains one or more rows/lines of [entity data](#entity-data-json-format), and is used as the "data file" when [building an index](how-to-index-build.md).
 
 > [!IMPORTANT]
-> While each **row** in an entity data file must be a valid JSON object, the overall file itself is not JSON but simply a *rows/lines of JSON objects*. The file contents are **NOT** meant to be an array of JSON objects.
+> Each **row** in an entity data file must be a valid JSON object, however the overall file content itself is not valid JSON object. The file contents are **NOT** meant to be an array of JSON objects, meaning files should not start/end with square brackets (```[]```) and lines should not end with a comma (```,```).
```
Comment (cyhuang01) [inline academic-services/knowledge-exploration-service/how-to-index-data.md:47]: >is not valid JSON object [](start = 107, length = 24)

is not a valid JSON object

File: academic-services/knowledge-exploration-service/how-to-index-data.md [hunk 2]
```diff
@@ -56,311 +56,291 @@ An entity data file contains one or more rows/lines of [entity data](#entity-dat
 The following entity JSON data is for an index that supports academic paper entities and uses a subset ([defined in the schema how-to](how-to-index-schema.md#example)) of the full [MAKES academic entity schema](reference-makes-api-entity-schema.md).
 
 > [!IMPORTANT]
-> The following example is purposefully using tabular formatting of the JSON data to make it easy to understand. However in practice **this data would have no such formatting and be contained on a *single* line**.
+> The following example shows each JSON attribute on a separate line to make it easier to read and understand for **documentation purposes only**. In the actual JSON data files to be used for indexing, MAKES requires each entities JSON to be contained **in a single line**.
```
Comment (cyhuang01) [inline academic-services/knowledge-exploration-service/how-to-index-data.md:59]: >each entities JSON [](start = 217, length = 18)

"each entities JSON" => "each entity's JSON"

File: academic-services/knowledge-exploration-service/how-to-grammar.md [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+# Grammar
+
+This document details the role of a grammar in allowing natural language processing in MAKES, the composition of a MAKES grammar, how to compile a MAKES grammar, and how to load a compiled MAKES grammar into a MAKES instance.
+
+## Natural language processing with grammars
+
+MAKES supports natural language query interpretation through the [Interpret API](reference-get-interpret.md), which requires a [Context Free Grammar (CFG)](https://academic.microsoft.com/topic/97212296) adhering to the [Speech Recognition Grammar Specification (SRGS)](https://www.w3.org/TR/speech-grammar/) format, a W3C standard for speech recognition grammars with support for generating semantic interpretations.
+
+In the context of the SRGS, the role of the MAKES grammar is as a *speech recognizer*. MAKES takes an input stream (natural language query) and tries to match it to a series of *rules* defined in the grammar, which in turn generate [structured query expressions](concept-query-expressions.md) as their output. Structured query expressions are what MAKES uses to retrieve entities matching a specific set of constraints.
```
Comment (cyhuang01) [inline academic-services/knowledge-exploration-service/how-to-grammar.md:16]: >concept-query-expressions [](start = 263, length = 25)

concepts-query-expressions

File: academic-services/knowledge-exploration-service/how-to-grammar.md [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+}
+```
+
+We can map the example rule expansion in the previous section to this response in the following ways:
+
+* "cumulative weight (log probability)" => "logprob"
+* "parse tree path" => "parse"
+* "top matching paper" => "rules"."output"."entities"[0]
+
+This response also highlights an important concept mentioned in the first part of this section, the generation of [structured query expressions](concept-query-expressions.md) as the output of each interpretation. These are shown in "rules"."output"."value".
```
Comment (cyhuang01) [inline academic-services/knowledge-exploration-service/how-to-grammar.md:285]: >concept-query-expressions.md [](start = 145, length = 28)

concepts-query-expressions.md

File: academic-services/knowledge-exploration-service/how-to-grammar.md (hunk 3 above)
Comment (cyhuang01) [inline academic-services/knowledge-exploration-service/how-to-grammar.md:16]: >concept-query-expressions.md [](start = 263, length = 28)

concepts-query-expressions.md

File: academic-services/knowledge-exploration-service/reference-grammar-syntax.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+The `tag` element specifies how a path through the grammar is to be interpreted.  It contains a sequence of semicolon-terminated statements.  A statement may be an assignment of a literal or a variable to another variable.  It may also assign the output of a function with 0 or more parameters to a variable.  Each function parameter may be specified using a literal or a variable.  If the function does not return any output, the assignment is omitted.  Variable scope is local to the containing rule.
+
+```xml
+<tag>x = 1; y = x;</tag>
+<tag>q = All(); q = And(q, q2);</tag>
+<tag>AssertEquals(x, 1);</tag>
+```
+
+Each `rule` in the grammar has a predefined variable named "out", representing the semantic output of the rule.  Its value is computed by evaluating each of the semantic statements traversed by the path through the `rule` matching the user query input.  The value assigned to the "out" variable at the end of the evaluation is the semantic output of the rule.  The semantic output of interpreting a user query against the grammar is the semantic output of the root rule.
```
Comment (cyhuang01) [inline academic-services/knowledge-exploration-service/reference-grammar-syntax.md:89]: >Each [](start = 0, length = 4)

How about adding "### out variable" before line 89?

**Review by cyhuang01 — CHANGES_REQUESTED** [review]
:clock1:

File: academic-services/knowledge-exploration-service/reference-grammar-syntax.md (hunk 5 above)
Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-grammar-syntax.md:89]: Great catch!

Comment (DarrinEide) [reply academic-services/knowledge-exploration-service/reference-grammar-syntax.md:89]: Restructured that general area a bit

**Review by cyhuang01 — APPROVED** [review]
:shipit:

File: academic-services/knowledge-exploration-service/reference-semantic-functions.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
+---
+title: Semantic functions
+description: Reference documentation for the semantic functions supported by the MAKES grammar
+ms.topic: reference
+ms.date: 9/1/2020
+---
+
+# Semantic functions
+
+Semantic functions associates semantic output with each interpreted path through the grammar.  In particular, the service evaluates the sequence of statements in the `tag` elements traversed by the interpretation to compute the final output.
```
Comment (RickRogahn) [inline academic-services/knowledge-exploration-service/reference-semantic-functions.md:10]: associate

**Review by RickRogahn — APPROVED** [review]
:shipit:

**Review by alch-msft — CHANGES_REQUESTED** [review]
Very detailed documentation. Good job

**Pull request merged**